File: android/app/src/main/kotlin/com/demo/grpcdemo/domain/exception/CatalogException.kt
package com.demo.grpcdemo.domain.exception

sealed class CatalogException(message: String) : Exception(message) {
    class NotFound(id: String)           : CatalogException("Product '$id' not found")
    class Unavailable(cause: Throwable?) : CatalogException("Service unavailable: ${cause?.message}")
    class Unauthenticated                : CatalogException("Session expired — please log in again")
    class Unknown(cause: Throwable?)     : CatalogException("Unknown error: ${cause?.message}")
}

File: flutter/lib/domain/exception/catalog_exception.dart
// flutter/lib/domain/exception/catalog_exception.dart

sealed class CatalogException implements Exception {
  const CatalogException(this.message);
  final String message;
  @override
  String toString() => message;
}

final class NotFoundException extends CatalogException {
  const NotFoundException(String id) : super("Product '$id' not found");
}

final class UnavailableException extends CatalogException {
  const UnavailableException(Object? cause)
      : super("Service unavailable: $cause");
}

final class UnknownException extends CatalogException {
  const UnknownException(Object? cause) : super("Unknown error: $cause");
}
